consensus/istanbul/proxy/proxy_handler.go

```diff
@@ -0,0 +1,352 @@
+					ph.sendValEnodeShareMsgs()
+					ph.sb.UpdateAnnounceVersion()
+				}
+				ph.sb.RemovePeer(proxy.node, p2p.ProxyPurpose)
+			}
+
+		case connectedPeer := <-ph.addProxyPeer:
+			// Proxied peer just connected.
+			// Set the corresponding proxyInfo's peer
+			peerNode := connectedPeer.Node()
+			peerID := peerNode.ID()
+			proxy := ph.ps.getProxy(peerID)
+			if proxy != nil {
+				logger.Debug("Connected proxy", "proxy", proxy.String(), "chan", "addProxyPeer")
+				if valsReassigned := ph.ps.setProxyPeer(peerID, connectedPeer); valsReassigned {
+					logger.Info("Remote validator to proxy assignment has changed.  Sending val enode share messages and updating announce version")
+					ph.sendValEnodeShareMsgs()
+					ph.sb.UpdateAnnounceVersion()
+				}
+
+				// Share this node's enodeCertificate for the proxy to use for handshakes via a forward message
+				proxyEnodeCertificateMsg, err := ph.sb.GenerateEnodeCertificateMsg(proxy.externalNode.URLv4())
+				if err != nil {
+					logger.Warn("Error generating enode certificate message", "err", err)
+				} else {
+					payload, err := proxyEnodeCertificateMsg.Payload()
+					if err != nil {
+						logger.Error("Error getting payload of enode certificate message", "err", err)
+					} else {
+						validatorAssignments := ph.ps.getValidatorAssignments(nil, []enode.ID{peerID})
+
+						destAddresses := make([]common.Address, 0, len(validatorAssignments))
+						for valAddress := range validatorAssignments {
+							destAddresses = append(destAddresses, valAddress)
+						}
+
+						// The forward message is being sent within a goroutine, since SendForwardMsg will query the proxy handler thread
+						// (which we are already in).  If this command is done inline, then it will result in a deadlock.
+						// TODO: Figure out a way to run this command inline.
+						go func() {
+							if err := ph.pe.SendForwardMsg(destAddresses, istanbul.EnodeCertificateMsg, nil, map[enode.ID][]byte{peerID: payload}); err != nil {
+								logger.Error("Error in forwarding a enodeCertificateMsg to proxy", "proxy", peerNode, "error", err)
+							}
+						}()
+					}
+				}
+			}
+
+		case disconnectedPeer := <-ph.removeProxyPeer:
+			// Proxied peer just disconnected.
+			peerID := disconnectedPeer.Node().ID()
+			if ph.ps.getProxy(peerID) != nil {
+				logger.Debug("Disconnected proxy peer", "peerID", peerID, "chan", "removeProxyPeer")
+				ph.ps.removeProxyPeer(peerID)
+			}
+
+		case proxyHandlerOp := <-ph.proxyHandlerOpCh:
+			proxyHandlerOp(ph.ps.getValidatorAssignments)
+			ph.proxyHandlerOpDoneCh <- struct{}{}
+
+		case <-ph.newBlockchainEpoch:
+			// New blockchain epoch. Update the validators in the proxySet
+			valsReassigned, error := ph.updateValidators()
+			if error != nil {
+				logger.Warn("Error in updating validator assignments on new epoch", "error", error)
+			}
+			if valsReassigned {
+				ph.sendValEnodeShareMsgs()
+				ph.sb.UpdateAnnounceVersion()
+			}
+
+		case <-phEpochTicker.C:
```

Comment:
what's the diff between `phEpochTicker` and `newBlockchainEpoch`?

Reply:
The proxy handler will wake up every `ph.proxyHandlerEpochLength` to do some "work".  E.g. resend out valenodeshare messages to the proxies (since it may have failed before).  So that's what `epoch` means for the proxy handler.  I know that it can be confusing since our `blockchain` also has epochs.  Can you think of a better name?

Reply:
mmm.. yeah, i would avoid `epoch` as a word here. 

Maybe `cycle`? or just `proxyTicker`?